src/AddThreepid.ts

```diff
@@ -67,23 +68,24 @@ export default class AddThreepid {
      * @param {string} emailAddress The email address to add
      * @return {Promise} Resolves when the email has been sent. Then call checkEmailLinkClicked().
      */
-    public addEmailAddress(emailAddress: string): Promise<IRequestTokenResponse> {
-        return MatrixClientPeg.get()
-            .requestAdd3pidEmailToken(emailAddress, this.clientSecret, 1)
-            .then(
-                (res) => {
-                    this.sessionId = res.sid;
-                    return res;
-                },
-                function (err) {
-                    if (err.errcode === "M_THREEPID_IN_USE") {
-                        err.message = _t("This email address is already in use");
-                    } else if (err.httpStatus) {
-                        err.message = err.message + ` (Status ${err.httpStatus})`;
-                    }
-                    throw err;
-                },
-            );
+    public async addEmailAddress(emailAddress: string): Promise<IRequestTokenResponse> {
+        try {
+            const res = await MatrixClientPeg.get().requestAdd3pidEmailToken(emailAddress, this.clientSecret, 1);
+            this.sessionId = res.sid;
+            return res;
+        } catch (err) {
+            if (err instanceof MatrixError && err.errcode === "M_THREEPID_IN_USE") {
+                throw new UserFriendlyError("This email address is already in use", { cause: err });
+            } else if (err instanceof HTTPError && err.httpStatus) {
+                throw new UserFriendlyError("%(errorMessage)s (Status %(httpStatus)s)", {
+                    errorMessage: err.message,
+                    httpStatus: err.httpStatus,
+                    cause: err,
+                });
```

Comment:
I don't think we need to re-label a `HTTPError` as they already have a good message that looks like `Server returned ${response.status} error: ${body}`

Reply:
Removed ⏩